src/typesystem.hpp

```diff
@@ -6,32 +6,23 @@
 #include <string>
 
 #include "typechecker_types.hpp"
+#include "unification.hpp"
 
 namespace Frontend {
 struct CompileTimeEnvironment;
 }
 
-enum class TypeFunctionTag { Var, Builtin, Sum, Product, Record };
-// A type function gives the 'real value' of a type.
-// This can refer to a sum type, a product type, a built-in type, etc.
-// For the purposes of the type system, we only care about the amount
-// of argument it takes.
-
-// If it's a var, equals points to another TypeFunctionHeader.
-// Else it points to a TypeFunctionData.
-struct TypeFunctionHeader {
-	TypeFunctionTag type;
-	int equals;
-};
-
+enum class TypeFunctionTag { Builtin, Sum, Product, Record };
 // Concrete type function. If it's a built-in, we use argument_count
 // to tell how many arguments it takes. Else, for sum, product and record,
 // we store their structure as a hash from names to monotypes.
 //
 // Dummy type functions are for unifying purposes only, but do not count
-// as 'deduced', because they were not created by the user
-
+// as 'deduced', because they were not created by the user/
+//
+// TODO: change for polymorphic approach
 struct TypeFunctionData {
+	TypeFunctionTag tag;
 	int argument_count; // -1 means variadic
 	std::unordered_map<std::string, MonoId> structure; // can be nullptr
```

Comment:
lol, did I write that `// can be nullptr` comment? I wonder what I was thinking?

Reply:
No, that was me actually... I realize now that I meant it can be _empty_ (when unused)